File: 0036-valid-sudoku/0036-valid-sudoku.py
"""
Time = O(n**2)
Space = O(n**2)
"""
from collections import defaultdict
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        
        rows = defaultdict(set)
        cols = defaultdict(set)
        boxes = defaultdict(set)
        
        for row in range(9):
            for col in range(9):
                cur = board[row][col]
                if cur != ".":
                    # Check row set
                    if cur in rows[row]:
                        return False
                    # Check col set
                    if cur in cols[col]:
                        return False
                    # Check box set
                    if cur in boxes[(row // 3, col // 3)]:
                        return False
                    rows[row].add(cur)
                    cols[col].add(cur)
                    boxes[(row // 3, col // 3)].add(cur)
                
        return True

File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        bool row[9][9] = {false};
        bool col[9][9] = {false};
        bool block[9][9] = {false};

        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                char cur = board[i][j];
                if (cur == '.') {
                    continue;
                }

                int idx = cur - '1';
                int area = (i / 3) * 3 + j / 3;

                if (row[i][idx] || col[j][idx] || block[area][idx]) {
                    return false;
                }

                row[i][idx] = true;
                col[j][idx] = true;
                block[area][idx] = true;
            }
        }
        
        return true;
    }
};